TombEngine/Objects/TR3/Entity/tr3_civvy.cpp

```diff
@@ -2,418 +2,601 @@
+		CIVVY_STATE_ATTACK_FAR_PUNCH = 8,
+		CIVVY_STATE_ATTACK_CLOSE_PUNCH = 9,
+		CIVVY_STATE_RUN = 10,
+		CIVVY_STATE_DEATH = 11,
+		CIVVY_STATE_CLIMB_2CLICKS = 12,
+		CIVVY_STATE_CLIMB_3CLICKS = 13,
+		CIVVY_STATE_CLIMB_4CLICKS = 14,
+		CIVVY_STATE_FALL_4CLICKS = 15
 	};
 
-	// TODO
 	enum CivvyAnim
 	{
+		CIVVY_ANIM_WALK = 0,
+		CIVVY_ANIM_WALKING_PUNCH_PREPARE = 1,
+		CIVVY_ANIM_WALKING_PUNCH_TO_IDLE = 2,
+		CIVVY_ANIM_WALKING_PUNCH_TO_WALK = 3,
+		CIVVY_ANIM_WALKING_PUNCH_ATTACK = 4,
+		CIVVY_ANIM_IDLE_TO_WALK = 5,
 		CIVVY_ANIM_IDLE = 6,
-
+		CIVVY_ANIM_WALKING_PUNCH_CANCEL = 7,
+		CIVVY_ANIM_WALK_TO_IDLE_LEFT = 8,
+		CIVVY_ANIM_CLOSE_PUNCH_CANCEL = 9,
+		CIVVY_ANIM_CLOSE_PUNCH_PREPARE = 10,
+		CIVVY_ANIM_FAR_PUNCH_CANCEL = 11,
+		CIVVY_ANIM_FAR_PUNCH_PREPARE = 12,
+		CIVVY_ANIM_CLOSE_PUNCH_TO_IDLE = 13,
+		CIVVY_ANIM_FAR_PUNCH_TO_IDLE = 14,
+		CIVVY_ANIM_FAR_PUNCH_TO_WALK = 15,
+		CIVVY_ANIM_CLOSE_PUNCH_ATTACK = 16,
+		CIVVY_ANIM_FAR_PUNCH_ATTACK = 17,
+		CIVVY_ANIM_IDLE_TO_WAIT = 18,
+		CIVVY_ANIM_WAIT_TO_IDLE = 19,
+		CIVVY_ANIM_WAIT = 20,
+		CIVVY_ANIM_RUN = 21,
+		CIVVY_ANIM_WALK_TO_RUN = 22,
+		CIVVY_ANIM_RUN_TO_WALK_LEFT = 23,
+		CIVVY_ANIM_RUN_TO_IDLE = 24,
+		CIVVY_ANIM_IDLE_TO_RUN = 25,
 		CIVVY_ANIM_DEATH = 26,
-		CIVVY_CLIMB3_ANIM = 27,
-		CIVVY_CLIMB1_ANIM = 28,
-		CIVVY_CLIMB2_ANIM = 29,
-		CIVVY_FALL4_ANIM = 30
+		CIVVY_ANIM_CLIMB_4CLICKS = 27,
+		CIVVY_ANIM_CLIMB_2CLICKS = 28,
+		CIVVY_ANIM_CLIMB_3CLICKS = 29,
+		CIVVY_ANIM_FALL_4CLICKS = 30,
+		CIVVY_ANIM_RUN_TO_WALK_RIGHT = 31,
 	};
 
 	void InitialiseCivvy(short itemNumber)
 	{
-		auto* item = &g_Level.Items[itemNumber];
+		auto& item = g_Level.Items[itemNumber];
 
 		InitialiseCreature(itemNumber);
-		SetAnimation(item, CIVVY_ANIM_IDLE);
+		SetAnimation(&item, CIVVY_ANIM_IDLE);
+	}
+
+
+	//In the future, if this function works, it could be taken to a more neutral place to be reused by other ally entities.
+	ItemInfo* CivvyFindNearTarget(short itemNumber, std::vector<GAME_OBJECT_ID>& excludedTargetList, int rangeDetection)
+	{
+		auto& item = g_Level.Items[itemNumber];
+		auto& creature = *GetCreatureInfo(&item);
+
+		ItemInfo* result = nullptr;
+
+		auto closestDistance = FLT_MAX;
+		float distanceValue;
+		float maxRange = rangeDetection <= 0 ? FLT_MAX : rangeDetection;
+
+		for (auto& targetCreature : ActiveCreatures)
+		{
+			//Ignore if it's itself, or a non valid Item.
+			if (targetCreature->ItemNumber == NO_ITEM || targetCreature->ItemNumber == itemNumber)
+				continue;
+
+			auto& currentItem = g_Level.Items[targetCreature->ItemNumber];
+
+			//Ignore if it's dead.
+			if (currentItem.HitPoints <= 0)
+				continue;
+
+			//Ignore if it's an entity from the Excluded Targets lists.
+			bool forbiddenTarget = false;
+			for (auto& excludedTargetID : excludedTargetList)
+			{
+				if (currentItem.ObjectNumber == excludedTargetID)
+				{
+					forbiddenTarget = true;
+					break;
+				}
+			}
+			if (forbiddenTarget)
```

Comment:
Please add empty new line before this one for more clean formatting.

Reply:
Done